### PR TITLE
Gem revamp

### PR DESCRIPTION
**Pull request by madtrick**
**Warning!!!** breaking changes ahead

New version of this middleman extension:

* This new version is no longer coupled to blog sites
* Custom mappers
* Rebuild only when changes in the imported data
* Entries stored as local data

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,72 +20,138 @@ Then as usual, run:
 bundle install
 ```
 
+## Usage
+
+Run `$middleman contentful` in your terminal. This will fetch entries for the configured
```
Comment (pxlpnk) [inline README.md]: using `$command_name` is not very nice to copy, we should remove the `$`

File: lib/contentful_middleman/commands/contentful.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+          contentful_instances = shared_instance.contentful_instances
+
+          contentful_instances.each do |instance|
+            old_version_hash = VersionHash.read_for_space(instance.space_name)
+
+            do_with_backup "#{instance.space_name}-data-backup", "data/#{instance.space_name}" do
+              entries =  instance.entries
+              entries.each do |entry|
+                context              = ContentfulMiddleman::Context.new
+                mapper               = instance.content_type_mapper entry.content_type.id
```
Comment (pxlpnk) [inline lib/contentful_middleman/commands/contentful.rb]: `()` missing around the function call

File: lib/contentful_middleman/commands/contentful.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      def self.source_root
+        ENV['MM_ROOT']
+      end
+
+      # Tell Thor to exit with a nonzero exit code on failure
+      def self.exit_on_failure?
+        true
+      end
+
+      def contentful
```
Comment (pxlpnk) [inline lib/contentful_middleman/commands/contentful.rb]: This is a pretty big method, I think we could extract some if it in its own methods

File: lib/contentful_middleman/commands/contentful.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+      private
+        def shared_instance
+          @shared_instance ||= ::Middleman::Application.server.inst
+        end
+
+        def data_file_path(instance, entry)
+          data_path(
+            instance.space_name,
+            instance.content_type_name(entry.content_type.id),
```
Comment (pxlpnk) [inline lib/contentful_middleman/commands/contentful.rb]: I am not a fan of method calls within a method call within a method call.

File: lib/contentful_middleman/commands/contentful.rb [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            entry.id)
+        end
+
+        def data_path(space_name, content_type_name, entry_id)
+          data_filename = "#{entry_id}.yaml"
+          File.join(
+            shared_instance.root_path.to_s,
+            'data',
```
Comment (pxlpnk) [inline lib/contentful_middleman/commands/contentful.rb]: magic number!

File: lib/contentful_middleman/commands/context.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
+module ContentfulMiddleman
+  class Context < BasicObject
+    def initialize
+      @_variables       = {}
```
Comment (pxlpnk) [inline lib/contentful_middleman/commands/context.rb]: Remove the `_` from the variables

File: lib/contentful_middleman/commands/context.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      @_nested_contexts = []
+    end
+
+    def method_missing(symbol, *args, &block)
+      if symbol =~ /.+=$/
```
Comment (pxlpnk) [inline lib/contentful_middleman/commands/context.rb]: you probably want to use `\A` and `\z` here.

File: lib/contentful_middleman/commands/context.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+      set field_name, new_contexts
+    end
+
+    def set(name, value)
+      @_variables[name.to_sym] = value
+    end
+
+    def get(name)
+      @_variables[name.to_sym]
```
Comment (pxlpnk) [inline lib/contentful_middleman/commands/context.rb]: maybe it makes sense to use `.fetch` here so you don't return `nil` when a key is not present and raise an error instead.

File: lib/contentful_middleman/commands/context.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+    def to_yaml
+      variables = @_variables.dup
+      variables.update(variables) do |k,v|
+        if @_nested_contexts.include? k
```
Comment (pxlpnk) [inline lib/contentful_middleman/commands/context.rb]: You should move this out in its own method

File: lib/contentful_middleman/core.rb [hunk 10]
```diff
@@ -1,69 +1,74 @@
...
-        access_token: options.access_token,
-        space: options.space,
-        dynamic_entries: :auto
-      )
-    end
+    def massage_content_types_options
+      content_types_options     = options.content_types
+      new_content_types_options = {}
+
+      content_types_options.each do |content_type_name, value|
```
Comment (pxlpnk) [inline lib/contentful_middleman/core.rb]: this looks like a job for `.each_with_object`

File: lib/contentful_middleman/instance.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    def content_type_name(content_type_id)
+      options.content_types.fetch(content_type_id).fetch(:name)
+    end
+
+    private
+    def client
+      @client ||= Contentful::Client.new(
+        access_token:     options.access_token,
+        space:            options.space.fetch(:id),
+        dynamic_entries:  :auto
```
Comment (pxlpnk) [inline lib/contentful_middleman/instance.rb]: You might want to set `raise_errors: true` so I fails loudly when something is breaking from the API side

File: lib/contentful_middleman/tools/backup.rb [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        end
+
+        def ensure_backup_path!
+          return if ::File.exists? basepath
+
+          FileUtils.mkdir_p basepath
+        end
+      end
+
+      ensure_backup_path!
```
Comment (pxlpnk) [inline lib/contentful_middleman/tools/backup.rb]: why not put this into the initialiser?

File: lib/contentful_middleman/tools/backup.rb [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+      def initialize(name, source)
+        @name   = name
+        @source = source
+
+        FileUtils.mkdir(path)
+        #FileUtils.cp(all_files_in_path(source), path)
```
Comment (pxlpnk) [inline lib/contentful_middleman/tools/backup.rb]: Remove the commented out code if not needed anymore.

File: lib/contentful_middleman/tools/backup.rb [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        def do_with_backup(backup_name, path_to_backup)
+          backup = create_backup backup_name,  path_to_backup
+
+          begin
+            yield
+          rescue StandardError => e
+            backup.restore
+            raise e
+          ensure
+            backup.destroy
```
Comment (pxlpnk) [inline lib/contentful_middleman/tools/backup.rb]: If something happens when restoring the backup (eg full disk, ..) we delete our backup without checking if the restore worked. Could this result in data loss?

File: README.md [hunk 15]
```diff
@@ -22,72 +20,138 @@ Then as usual, run:
...
-  f.space = 'YOUR_SPACE_ID'
+  f.space         = SPACE
+  f.access_token  = ACCESS_TOKEN
+  f.cda_query     = QUERY
+  f.content_types = CONTENT_TYPES_MAPPINGS
+end
+```
+
+Parameter | Description
+----------|------------
```
Comment (pxlpnk) [inline README.md]: the table is very hard to read in plain text

File: lib/contentful_middleman/instance.rb [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        mapper_class.new
+      end
+    end
+
+    def space_name
+      @space_name ||= options.space.fetch(:name)
+    end
+
+    def content_type_name(content_type_id)
+      options.content_types.fetch(content_type_id).fetch(:name)
```
Comment (pxlpnk) [inline lib/contentful_middleman/instance.rb]: this should be split up to

**Comment by pxlpnk** [comment]
Looks good so far, can you also sprinkle some `(` and `)` around the method calls :cherry_blossom:

**Comment by pxlpnk** [comment]
I guess @asaaki will have a lot more to add

File: README.md (hunk 1 above)
Comment (madtrick) [inline README.md]: Changed in https://github.com/madtrick/contentful_middleman/commit/0317577bb553e95291640fcb26b9c42470578a1f

File: lib/contentful_middleman/commands/contentful.rb (hunk 3 above)
Comment (madtrick) [inline lib/contentful_middleman/commands/contentful.rb]: Fixed in https://github.com/madtrick/contentful_middleman/commit/2a2518156f5545a2972935906f2533e5e45371f0

File: lib/contentful_middleman/commands/contentful.rb (hunk 5 above)
Comment (madtrick) [inline lib/contentful_middleman/commands/contentful.rb]: Fixed here https://github.com/madtrick/contentful_middleman/commit/2a2518156f5545a2972935906f2533e5e45371f0#diff-13ed9bdaf79c345622833edc8a5e858aR20

File: lib/contentful_middleman/commands/context.rb (hunk 6 above)
Comment (madtrick) [inline lib/contentful_middleman/commands/context.rb]: Fixed here https://github.com/madtrick/contentful_middleman/commit/31d67fbc9f4cd4efa48239c4bdf678f9ac29dab0

File: lib/contentful_middleman/commands/context.rb (hunk 7 above)
Comment (madtrick) [inline lib/contentful_middleman/commands/context.rb]: Fixed here https://github.com/madtrick/contentful_middleman/commit/92effa8fed44546eddcd5a96cc37577b9a60c857

File: lib/contentful_middleman/commands/context.rb (hunk 8 above)
Comment (madtrick) [inline lib/contentful_middleman/commands/context.rb]: Fixed here https://github.com/madtrick/contentful_middleman/commit/a4d973297b83f4bfb9fab961f9a238f61d59fa90

File: lib/contentful_middleman/commands/context.rb (hunk 9 above)
Comment (madtrick) [inline lib/contentful_middleman/commands/context.rb]: Fixed here https://github.com/madtrick/contentful_middleman/commit/a4d973297b83f4bfb9fab961f9a238f61d59fa90

File: lib/contentful_middleman/core.rb (hunk 10 above)
Comment (madtrick) [inline lib/contentful_middleman/core.rb]: Done https://github.com/madtrick/contentful_middleman/commit/749f5d459346a8c1349c74361de44c8300a7e6cd

File: lib/contentful_middleman/instance.rb (hunk 11 above)
Comment (madtrick) [inline lib/contentful_middleman/instance.rb]: Done https://github.com/madtrick/contentful_middleman/commit/fa233208f04a205cdc4e310d616db1f2a6bcb1d6

File: lib/contentful_middleman/tools/backup.rb (hunk 12 above)
Comment (madtrick) [inline lib/contentful_middleman/tools/backup.rb]: Don't why I did that. Changed here https://github.com/madtrick/contentful_middleman/commit/67f874d328f7c3b8e9daba91396359074c48110b

File: lib/contentful_middleman/tools/backup.rb (hunk 14 above)
Comment (madtrick) [inline lib/contentful_middleman/tools/backup.rb]: Now we delete the backup only if there was no problem while executing the block or if the backup has been properly restored. Done here https://github.com/madtrick/contentful_middleman/commit/9ff953d0d3b97ae1f42e39beee0facfc383e3542

It isn't a bulletproof solution but we aren't creating here a backup solution 😝

File: README.md (hunk 15 above)
Comment (madtrick) [inline README.md]: Done here https://github.com/madtrick/contentful_middleman/commit/598aef2daa4e3049312a3cd9f2df6c8d7b3ca806 my dear

File: lib/contentful_middleman/instance.rb (hunk 16 above)
Comment (madtrick) [inline lib/contentful_middleman/instance.rb]: This code is gone

**Comment by asaaki** [comment]
We should remove the version bump from this PR.
(It's quite common to do this in a dedicated commit/PR to avoid possible conflicts with concurrent PRs.)

- Remove bump in `version.rb`
- Adjust changelog headline to `[Unreleased][unreleased]` (like proposed at <http://keepachangelog.com/>).

Will review the rest of the PR later.

**Comment by madtrick** [comment]
@asaaki good point! Done.

**Comment by pxlpnk** [comment]
Looks good to me now, should we make one commit out if this?

**Comment by madtrick** [comment]
Yes, I think squashing all this mess makes sense

**Comment by madtrick** [comment]
I'll wait for @asaaki comments before doing that

**Comment by asaaki** [comment]
I don't mind, if you really want to get it out now.
Otherwise I would have a final look at this PR tonight.
Also: Some tweaks can always follow in upcoming PR(s).
It's up to you to decide.

**Comment by madtrick** [comment]
@asaaki so let's get this merged and fix anything else in upcoming commits

**Comment by asaaki** [comment]
@pxlpnk has to do this, I'm not an official collaborator/maintainer. :wink:

**Comment by pxlpnk** [comment]
:+1:  on this.
If you could add a note about the breaking changes to the changelog (I know a major release kind of implies this but better be too verbose here) and squash the commits.

**Comment by madtrick** [comment]
@pxlpnk whenever you want

**Comment by pxlpnk** [comment]
Merging it somewhat later today and will release a new version on rubygems.

**Pull request merged**